bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java

```diff
@@ -0,0 +1,578 @@
+        }
+
+        // This has been pushed into this method to defer initialisation as this makes testing easier.
+        protected void loadFileInfo() {
+            if (fBndFile != null) {
+                return;
+            }
+            final ICompilationUnit compUnit = fContext.getCompilationUnit();
+            final IJavaProject java = compUnit.getJavaProject();
+            final IProject eclipse = java.getProject();
+            fBndFile = eclipse.getFile(Project.BNDFILE);
+        }
+
+        protected Workspace getWorkspace() throws Exception {
+            return Central.getWorkspace();
+        }
+
+        protected void loadModel() throws CoreException {
+            if (fBuildPathVersioned != null) {
+                return;
+            }
+            loadFileInfo();
+            String contents;
+            try {
+                contents = IO.collect(new InputStreamReader(fBndFile.getContents(), fBndFile.getCharset()));
+                fBndDoc = new Document(contents);
+                fBndModel = new BndEditModel(getWorkspace());
+                fBndModel.loadFrom(fBndDoc);
+            } catch (IOException e) {
+                throw new CoreException(new Status(IStatus.ERROR, BndtoolsConstants.CORE_PLUGIN_ID, "Error trying to read " + fBndFile, e));
+            } catch (Exception e) {
+                throw new CoreException(new Status(IStatus.ERROR, BndtoolsConstants.CORE_PLUGIN_ID, "Error trying to access Bnd workspace", e));
+            }
+
+            final List<VersionedClause> origBuildPath = fBndModel.getBuildPath();
+            fBuildPath = origBuildPath == null ? new ArrayList<>() : new ArrayList<>(origBuildPath);
+            final int capacity = 2 * fBuildPath.size() + 1;
+            fBuildPathBundles = new HashSet<>(capacity);
+            fBuildPathVersioned = new HashSet<>(capacity);
+            for (VersionedClause bundleVersion : fBuildPath) {
+                fBuildPathBundles.add(bundleVersion.getName());
+                fBuildPathVersioned.add(bundleVersion);
+            }
+        }
+
+        public List<VersionedClause> getBuildPath() throws CoreException {
+            loadModel();
+            return fBuildPath;
+        }
+
+        public IFile getBndFile() {
+            loadFileInfo();
+            return fBndFile;
+        }
+
+        public void addBundle(String bundle) throws CoreException {
+            addBundle(new VersionedClause(bundle, new Attrs()));
+        }
+
+        public void addBundle(VersionedClause versionedBundle) throws CoreException {
+            loadModel();
+            fBuildPathBundles.add(versionedBundle.getName());
+            if (fBuildPathVersioned.add(versionedBundle)) {
+                fBuildPath.add(versionedBundle);
+            }
+            fBndModel.setBuildPath(fBuildPath);
+            fBndModel.saveChangesTo(fBndDoc);
+            InputStream str;
+            try {
+                str = new ByteArrayInputStream(fBndDoc.get()
```

Comment:
In general, you want to avoid ByteArray{Input,Output}Stream classes. They are excessively synchronized. We have ByteBuffer{Input,Output}Stream in the aQute.lib.io package as replacements.

Reply:
Changed to use `ByteBufferInputStream`.